core/shared/src/main/scala/chartreuse/Plot.scala

```diff
@@ -50,19 +52,197 @@ final case class Plot[
     copy(yTitle = newYTitle)
   }
 
-  //  TODO: handle axes and grid layout
-  def draw(width: Int, height: Int): Picture[
-    Alg & Text & doodle.algebra.Transform & Debug,
-    Unit
-  ] = {
-    val plot =
+  def withGrid(newGrid: Boolean): Plot[A, Alg] = {
+    copy(grid = newGrid)
+  }
+
+  def draw(width: Int, height: Int): Picture[Alg, Unit] = {
+    val allData = layers.flatMap(_.data.foldLeft(List.empty[A])(_ :+ _))
+    val dataBoundingBox = Data(allData).boundingBox(layers.head.toPoint)
+
+    val minX = dataBoundingBox.left
+    val maxX = dataBoundingBox.right
+    val minY = dataBoundingBox.bottom
+    val maxY = dataBoundingBox.top
+
+    val scale = Scale.linear.build(dataBoundingBox, width, height)
+
+    val xTicks = TickMarkCalculator.calculateTickScale(minX, maxX, 12)
+    val yTicks = TickMarkCalculator.calculateTickScale(minY, maxY, 12)
+
+    // Map the Ticks to the screen coordinates
+    val xTicksMapped = Ticks(
+      scale(Point(xTicks.min, 0)).x,
+      scale(Point(xTicks.max, 0)).x,
+      xTicks.size
+    )
+    val yTicksMapped = Ticks(
+      scale(Point(0, yTicks.min)).y,
+      scale(Point(0, yTicks.max)).y,
+      yTicks.size
+    )
+
+    // Convert the Ticks to a sequence of points
+    val xTicksSequence = xTicksToSequence(xTicks, scale)
+    val yTicksSequence = yTicksToSequence(yTicks, scale)
+
+    val allLayers =
       layers
         .map(_.draw(width, height))
         .foldLeft(empty[Alg])(_ on _)
-        .margin(20)
-        .debug(color = Color.black)
-        .margin(5)
 
+    val plotWithXTicks = withXTicks(xTicksSequence, allLayers, yTicksMapped)
+    val plotWithXAndYTicks =
+      withYTicks(yTicksSequence, plotWithXTicks, xTicksMapped)
+    val plotWithTicksAndAxes =
+      withAxes(plotWithXAndYTicks, xTicksMapped, yTicksMapped)
+
+    if (grid) {
+      val plotWithTicksAndAxesAndGrid = withGrid(
+        plotWithTicksAndAxes,
+        xTicksMapped,
+        yTicksMapped,
+        xTicksSequence,
+        yTicksSequence
+      )
+      return withTitles(plotWithTicksAndAxesAndGrid)
+    }
+
+    withTitles(plotWithTicksAndAxes)
+  }
+
+  /** Converts `Ticks` to a list of tuples. The first element is the mapped
+    * coordinate of a tick - to place the tick on a graph. The second one is the
+    * original coordinate - to give the tick a label with its coordinate.
+    */
+  private def xTicksToSequence(
```

Comment:
Too much repetition in this code.  There  shouldn't be repeated code for the x- and y-axis. You can abstract over x- and y-coordinates using a function. E.g. 

```scala
val asX: Double => Point = x => Point(x, 0)
val asY: Double => Point = y => Point(0, y)
```